src/crimson/admin/admin_socket.h

```diff
@@ -0,0 +1,320 @@
+
+  std::unique_ptr<AdminSocketHook> version_hook;
+  std::unique_ptr<AdminSocketHook> git_ver_hook;
+  std::unique_ptr<AdminSocketHook> the0_hook;
+  std::unique_ptr<AdminSocketHook> help_hook;
+  std::unique_ptr<AdminSocketHook> getdescs_hook;
+  std::unique_ptr<AdminSocketHook> test_throw_hook;  // for dev unit-tests
+
+  struct parsed_command_t {
+    std::string            m_cmd;
+    cmdmap_t               m_parameters;
+    std::string            m_format;
+    const AsokServiceDef*  m_api;
+    const AdminSocketHook* m_hook;
+    ::seastar::gate*       m_gate;      //!< the gate of the server-block where the command was located
+    /*!
+        the length of the whole command-sequence under the 'prefix' header
+     */
+    std::size_t            m_cmd_seq_len;
+  };
+
+  /*!
+    parse the incoming command line into the sequence of words that identifies the API,
+    and into its arguments.
+    Locate the command string in the registered blocks.
+  */
+  seastar::future<std::optional<AdminSocket::parsed_command_t>> parse_cmd(const std::string command_text);
+
+  struct server_block {
+    server_block(const std::vector<AsokServiceDef>& hooks)
+      : m_hooks{hooks}
+    {}
+    const std::vector<AsokServiceDef>& m_hooks;
+    mutable ::seastar::gate m_gate;
+  };
+
+  // \todo possible improvement: cache all available commands, from all servers, in one vector.
+  //  Recreate the list every register/unreg request.
+
+  /*!
+    The servers table is protected by a rw-lock, to be acquired exclusively only
+    when registering or removing a server.
+    Note that the lock is *not* held when executing a specific hook. As the map
+    keeps stable item addresses, we do not worry about a specific API block
+    disappearing from under our feet (each individual block is protected from removal
+    by a seastar::gate). Still, we must guarantee the ability to safely perform
+    map lookups.
+   */
+  seastar::shared_mutex servers_tbl_rwlock;
+  std::map<hook_server_tag, server_block> servers;
+
+  struct GateAndHook {
+    ::seastar::gate* m_gate;
+    const AsokServiceDef* api;
+  };
+
+  /*!
+    locate_command() will search all servers' control blocks. If found, the
+    relevant gate is entered. Returns the AsokServiceDef, and the "activated" gate.
+   */
+  AdminSocket::GateAndHook locate_command(std::string_view cmd);
+
+  /*!
+    Find the longest subset of words in 'match' that is a registered API (in any of the
+    servers' control blocks).
+    The search is conducted with the servers table RW-lock held.
+    If a matching API is found, the
+    relevant gate is entered. Returns the AsokServiceDef, and the "activated" gate.
+
+    Note: uses an async seastar::thread to wait for servers_tbl_rwlock. The returned
+    future is guaranteed to be available.
+   */
+  seastar::future<AdminSocket::GateAndHook> locate_subcmd(std::string match);
+
+public:
+  /*!
+    iterator support
+   */
+  AdminHooksIter begin();
```

Comment:
I think I'm missing something, where's the user for this?

Reply:
Every range-based for loop, e.g. in HelpHook::exec_command():
  for (const auto& hk_info : *m_as) {..}